pkg/ingester/downscale.go

```diff
@@ -0,0 +1,65 @@
+package ingester
+
+import (
+	"net/http"
+
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/services"
+
+	"github.com/grafana/mimir/pkg/util"
+)
+
+// PrepareInstanceRingDownscaleHandler prepares the ingester ring entry for downscaling. It can mark ingester as READONLY or back to ACTIVE.
+//
+// Following methods are supported:
+//
+//   - GET
+//     Returns timestamp when instance was switched to READONLY state, or 0, if partition is not in READONLY state.
+//
+//   - POST
+//     Switches the instance to READONLY state (if not yet), and returns the timestamp when the switch to
+//     READONLY state happened.
+//
+//   - DELETE
+//     Sets partition back from READONLY to ACTIVE state.
+func (i *Ingester) PrepareInstanceRingDownscaleHandler(w http.ResponseWriter, r *http.Request) {
```

Comment:
Shall we add a unit test for this handler?

Reply:
I started in https://github.com/grafana/mimir/pull/8956/commits/9bb33466be1ee28abebd65fcd91f4c3cfb1fd328, but it needs more work.

Reply:
Added some more test cases